PorousFlowNearestNodeNumber Material added

WilkAndy opened the pull request:
### Complete each of the following items before creating a PR

Hi @andrsd ,

This is not ready to merge (all the porous_flow tests will fail anyway) but I'm hoping you can glance over PorousFlowNearestNodeNumber.

For each qp, it stores the nearest node number in the element.  It'll get used by all the other PorousFlow Materials so they can store the correct nodal values at the associated nearest quadpoint.  Then those nodal values can then be used by Kernels, BCs, etc, straightforwardly.

Perhaps the only strange thing is the _on_initial_only flag - i wanted something that would mean the nearest-node computation was only performed once (for non-adapting meshes).  I think this does it correctly, but am a bit disappointed to need Old values, and maybe the copying of current to Old is expensive.  I didn't want to spend too long on this in case you can see something obviously stupid.

Refs #7333
Refs #7334

andrsd commented on modules/porous_flow/src/materials/PorousFlowNearestNodeNumber.C:
```diff
@@ -0,0 +1,56 @@
+{
+  unsigned n = 0;
+  Real smallest_dist = (_current_elem->point(0) - _q_point[_qp]).size();
+  for (unsigned i = 1; i < _current_elem->n_nodes(); ++i)
```
I belive nodes are indexed from 0.

andrsd replied:
Never mind, I see what you are doing... this is correct...

andrsd commented:
This looks good. If you want to use the stateful material props to pre-compute the nearest indices, keep in mind that every time step MOOSE will shift your current props to old ones. So, in the next step you have to initialize the current values of your material props. But, you can just assign the old one into the current one, so you do not have to re-compute every time step.

WilkAndy commented:
This is still "Under Development".  I have implemented the "NodeNumber" stuff (#7334), but still need to implement the BC stuff that I originally wanted to do (#7333).

WilkAndy committed:
PorousFlowNodeNumber and PorousFlow BCs added.

PorousFlow Materials store nodal information (such as porepressures, densities, etc) at quadpoints.  This Material contains the map that defines which quadpoint stores which node's information.
_node_number[qp] = the element's node number that is stored at quadpoint=qp

This Material also checks that the number of nodes does not exceed the number of quadpoints in each element, for elements not on a boundary.  If the check fails then clearly the other PorousFlow Materials won't work (there won't be enough quadpoints to store all the required nodal information).

Fixes #7334

There are 4 BCs.
The basic one uses a Function to specify the flux on the boundary, and also has the option of multiplying by the fluid mobility, the relative permeability and/or mass fractions, which are all useful in the case of sinks to prevent an unlimited amount of fluid being withdrawn from the porous medium.
Derived from the basic one, is another BC that allows the flux to be modified by a piecewise-linear function of porepressure, which is useful for the case where transfer coefficients are defined across the boundary, or more complicated situations.
Derived from the basic one are two others, in which the flux is governed by a half Gaussian or half cubic function of porepressure, which are useful for modelling evapotranspiration through a boundary.

Fixes #7333

WilkAndy committed:
Existing PorousFlow Materials now use PorousFlowNodeNumber.

Fixes #7334

WilkAndy committed:
Tests of the BCs, and doco of these tests.

All other tests modified to use PorousFlowNodeNumber.

Addressed all of David's comments.

Fixes #7333
Fixes #7334

WilkAndy commented:
This PR is ready for inspection.  @cpgr will be interested in this.

@cpgr: I need to write more tests and doco for the BCs, but the 9 tests i've included are OK for an initial start, and much of the code is copied directly from the richards module so it's likely to be too wrong.   I'm ready to go home for the weekend, so will write the extra tests and doco as part of another PR.

WilkAndy commented:
Is this PR passable, @andrsd ?

andrsd commented on modules/porous_flow/include/bcs/PorousFlowHalfCubicSink.h:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Applies a flux sink to a boundary.  The base flux
+ * defined by PorousFlowSink is multiplied by a cubic.
+ * Denote x = porepressure - centre.  Then
```
Here in 'merica, they spell `centre` as `center`. I do not really care too much, but if somebody feels strongly about consistent English throughout the code, `center` should be used... Not just here, but in the rest of the code as well.

WilkAndy replied:
fixed

andrsd commented on modules/porous_flow/src/bcs/PorousFlowSink.C:
```diff
@@ -0,0 +1,179 @@
+    _mass_fractions(getMaterialProperty<std::vector<std::vector<Real> > >("PorousFlow_mass_frac")),
+    _dmass_fractions_dvar(getMaterialProperty<std::vector<std::vector<std::vector<Real> > > >("dPorousFlow_mass_frac_dvar")),
+    _node_number(getMaterialProperty<unsigned int>("PorousFlow_node_number")),
+    _qp_map(0)
```
This can be a little bit confusing. The default c-tor should do the right thing, i.e. you would not list `_qp_map` here at all...

WilkAndy replied:
fixed

andrsd commented on modules/porous_flow/tests/jacobian/hgs01.i:
```diff
@@ -0,0 +1,232 @@
+  [./check]
+    type = SMP
+    full = true
+    #petsc_options = '-snes_test_display'
```
remove this line

WilkAndy replied:
fixed

andrsd commented on modules/porous_flow/tests/jacobian/pls03.i:
```diff
@@ -0,0 +1,274 @@
+  [./check]
+    type = SMP
+    full = true
+    #petsc_options = '-snes_test_display'
```
remove this line

WilkAndy replied:
fixed

andrsd commented on modules/porous_flow/tests/jacobian/pls02.i:
```diff
@@ -0,0 +1,227 @@
+  [./check]
+    type = SMP
+    full = true
+    #petsc_options = '-snes_test_display'
```
remove this line

WilkAndy replied:
fixed

andrsd commented on modules/porous_flow/tests/jacobian/hcs01.i:
```diff
@@ -0,0 +1,232 @@
+  [./check]
+    type = SMP
+    full = true
+    #petsc_options = '-snes_test_display'
```
remove

WilkAndy replied:
fixed

andrsd commented:
The code is fine, cannot comment on the physics. After fixing the minor issues this is ready to merge.

WilkAndy commented:
Thanks @andrsd .  I have addressed all your comments.  Also, i have now fully tested and documented the code, so it can all go into one merge, which is nice.

Two of the tests cannot yet run in debug mode unfortunately, because of the issue #6774 .   Opt mode is fine, so i'm happy enough that the current code is correctly implementing the sinks that i want.   I'm hoping that issue #6774 will be addressed soon, nudge nudge on @aeslaughter :stuck_out_tongue:

WilkAndy commented:
Thanks @andrsd ,

a
